File: internal/bot/bot.go
package bot

import (
	tgbotapi "github.com/go-telegram-bot-api/telegram-bot-api/v5"
	"github.com/konstunn/cy-drive-exam-slot-finder/pkg"
)

// Bot represents a Telegram bot for finding driving exam slots
type Bot struct {
	api     pkg.TelegramBot
	scraper pkg.Scraper
}

// NewBot creates a new bot instance
func NewBot(api pkg.TelegramBot, scraper pkg.Scraper) *Bot {
	return &Bot{
		api:     api,
		scraper: scraper,
	}
}

// getReply generates a reply based on the incoming message text
func (b *Bot) getReply(text string) string {
	switch text {
	case "/start":
		return "Hello! I'm a driving exam slot finder bot. Send me any text, and I'll repeat it."
	case "/find_slots":
		return "Searching for available slots..."
	default:
		return "You said: " + text
	}
}

// HandleUpdate processes a single Telegram update
func (b *Bot) HandleUpdate(update tgbotapi.Update) {
	if update.Message == nil {
		return
	}

	reply := b.getReply(update.Message.Text)

	msg := tgbotapi.NewMessage(update.Message.Chat.ID, reply)
	msg.ReplyToMessageID = update.Message.MessageID

	b.api.Send(msg)
}

// Run starts the bot's main event loop
func (b *Bot) Run() {
	u := tgbotapi.NewUpdate(0)
	u.Timeout = 60

	updates := b.api.GetUpdatesChan(u)

	for update := range updates {
		b.HandleUpdate(update)
	}
}

File: src/bot.js
const TelegramBot = require('node-telegram-bot-api');

/**
 * Bot represents a Telegram bot for finding driving exam slots
 */
class Bot {
  /**
   * @param {string} token - Telegram bot token
   * @param {ChromeScraper} scraper - The scraper instance
   */
  constructor(token, scraper) {
    this.bot = new TelegramBot(token, { polling: true });
    this.scraper = scraper;

    this.setupEventHandlers();
  }

  /**
   * Generate a reply based on the incoming message text
   * @param {string} text
   * @returns {string}
   */
  getReply(text) {
    switch (text) {
      case '/start':
        return 'Hello! I\'m a driving exam slot finder bot. Send me any text, and I\'ll repeat it.';
      case '/find_slots':
        return 'Searching for available slots...';
      default:
        return `You said: ${text}`;
    }
  }

  /**
   * Set up event handlers for the bot
   */
  setupEventHandlers() {
    this.bot.on('message', (msg) => {
      this.handleMessage(msg);
    });
  }

  /**
   * Handle incoming messages
   * @param {Object} msg - Telegram message object
   */
  async handleMessage(msg) {
    const chatId = msg.chat.id;
    const text = msg.text;

    const reply = this.getReply(text);

    try {
      await this.bot.sendMessage(chatId, reply, {
        reply_to_message_id: msg.message_id
      });
    } catch (error) {
      console.error('Error sending message:', error);
    }
  }

  /**
   * Start the bot
   */
  start() {
    console.log('Bot started and listening for messages...');
  }

  /**
   * Stop the bot
   */
  stop() {
    this.bot.stopPolling();
  }
}

module.exports = Bot;